cmd/pj-rehearse/main.go

```diff
@@ -1,23 +1,278 @@
+// are modified to use this ConfigMap instead of the "production" one.
+
+var ciOperatorConfigsCMName = "ci-operator-configs"
+
+type rehearsalCIOperatorConfigs struct {
+	cmclient  corev1.ConfigMapInterface
+	prNumber  int
+	configDir string
+
+	logger logrus.FieldLogger
+	dry    bool
+
+	configMapName string
+	neededConfigs map[string]string
+}
+
+func newRehearsalCIOperatorConfigs(cmclient corev1.ConfigMapInterface, prNumber int, configDir string, logger logrus.FieldLogger, dry bool) *rehearsalCIOperatorConfigs {
+	name := fmt.Sprintf("rehearsal-ci-operator-configs-%d", prNumber)
+	return &rehearsalCIOperatorConfigs{
+		cmclient:      cmclient,
+		prNumber:      prNumber,
+		configDir:     configDir,
+		logger:        logger.WithField("ciop-configs-cm", name),
+		dry:           dry,
+		configMapName: name,
+		neededConfigs: map[string]string{},
+	}
+}
+
+// If a job uses the `ci-operator-config` ConfigMap, save which key does it use
+// from it and replace that ConfigMap reference with a reference to the
+// temporary, rehearsal ConfigMap containing the necessary keys with content
+// matching the version from tested PR
+func (c *rehearsalCIOperatorConfigs) FixupJob(job *prowconfig.Presubmit, repo string) {
+	for _, container := range job.Spec.Containers {
+		for _, env := range container.Env {
+			if env.ValueFrom == nil {
+				continue
+			}
+			if env.ValueFrom.ConfigMapKeyRef == nil {
+				continue
+			}
+			if env.ValueFrom.ConfigMapKeyRef.Name == ciOperatorConfigsCMName {
+				filename := env.ValueFrom.ConfigMapKeyRef.Key
+				env.ValueFrom.ConfigMapKeyRef.Name = c.configMapName
+				c.neededConfigs[filename] = filepath.Join(repo, filename)
+
+				logFields := logrus.Fields{"ci-operator-config": filename, "rehearsal-job": job.Name}
+				c.logger.WithFields(logFields).Info("Rehearsal job uses ci-operator config ConfigMap")
+			}
+		}
+	}
+}
+
+// Create a rehearsal ConfigMap with ci-operator config files needed by the
+// rehearsal jobs.
+func (c *rehearsalCIOperatorConfigs) Create() error {
+	cm := &v1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{Name: c.configMapName},
+		Data:       map[string]string{},
+	}
+	c.logger.Debug("Preparing rehearsal ConfigMap for ci-operator configs")
+
+	for key, path := range c.neededConfigs {
+		fullPath := filepath.Join(c.configDir, path)
+		content, err := ioutil.ReadFile(fullPath)
+		c.logger.WithField("ciop-config", key).Info("Loading ci-operator config to rehearsal ConfigMap")
+		if err != nil {
+			return fmt.Errorf("failed to read ci-operator config file from %s: %v", fullPath, err)
+		}
+
+		cm.Data[key] = string(content)
+	}
+
+	if c.dry {
+		cmAsYAML, err := yaml.Marshal(cm)
+		if err != nil {
+			return fmt.Errorf("Failed to marshal ConfigMap to YAML: %v", err)
+		}
+		fmt.Printf("%s\n", cmAsYAML)
+		return nil
+	}
+	c.logger.Info("Creating rehearsal ConfigMap for ci-operator configs")
+	_, err := c.cmclient.Create(cm)
```

Comment:
Do we have any cleanup yet?

Reply:
Not yet, current version just spawns the jobs and exits, it cannot clean up these CMs at that moment because some ProwJob-created Pods might not even be running yet. And waiting on jobs to finish is @bbguimaraes' turf.

We'll eventually want to add some cleanup even to the startup, but that's an increment.